src/Compilers/CSharp/Portable/Symbols/Source/SourcePropertySymbolBase.cs

```diff
@@ -1349,6 +1415,11 @@ internal override void ForceComplete(SourceLocation locationOpt, CancellationTok
                         GetAttributes();
                         break;
 
+                    case CompletionPart.StartPropertyEnsureSignature:
+                    case CompletionPart.FinishPropertyEnsureSignature:
+                        EnsureSignature();
```

Comment:
Believe we need the following here

```c#
_state.SpinWaitComplete(CompletionPart.FinishPropertyEnsureSignature, cancellationToken);
```

This is necessary if the call to `EnsureSignature` could call out to `ForceComplete`. In that case the call to `EnsureSignature` would be re-entrant here and hence wouldn't satisfy the contract of `ForceComplete`. 

Unsure if this is or isn't possible given the call graph of `EnsureSignature`. At the least I think we should have an assert that validates the part is marked as complete after `EnsureSignature` is called.

Reply:
We have SpinWaitComplete after this switch.

> if the call to EnsureSignature could call out to ForceComplete. 

I do not think this can happen. ForceComplete is there to make sure we collected all declaration diagnostics, it is not meant to be used for any other purpose.

I will add the suggested assert. 

---
In reply to: [525326274](https://github.com/dotnet/roslyn/pull/49360#discussion_r525326274) [](ancestors = 525326274)